### DISCUSSION TITLE
Overriding forms to submit via AJAX, but AJAX success event is not dispatched

### DISCUSSION THREAD

**Discussion by chrismlusk** [solspace/craft-freeform#460] [Questions & Answers] [answered]
Basically, when a form's success behavior is "redirect-return-url" in the form builder but you programmatically enable AJAX, Freeform does not emit all of the AJAX events.

I am forcing all forms to submit via AJAX by modifying the `options` object like so:

```js
form.addEventListener('freeform-ready', (event) => {
  event.options.ajax = true;
});
```

Then, I want to perform various operations after a form has been successfully posted. But, the order of operations in the plugin's `_onSubmitAjax()` method prevent the `freeform-ajax-success` event from ever being dispatched. ([See relevant lines here](https://github.com/solspace/craft-freeform/blob/master/packages/scripts/src/components/front-end/plugin/freeform.js#L679-L708).)

```js
_onSubmitAjax = (event) => {
  // ...

  if (finished && response.onSuccess === SUCCESS_BEHAVIOUR_REDIRECT_RETURN_URL && returnUrl) {
    window.location.href = returnUrl;
    return;
  }

  // ...

  this._dispatchEvent(EventTypes.EVENT_AJAX_SUCCESS, { request, response });
};
```

I could update all forms in the form builder to use a different success behavior, but that would be obscure and error prone. Besides, I want the client to continue building forms using whichever success behavior they'd like — I just need to execute code for various integrations/features only when a form submission is successful.

If there were a way I could override the success behavior before the request is made, my forms wouldn't meet that conditional shown above and then I redirect the user myself in the `freeform-ajax-submit` event callback. 

I couldn't find a way to do this, though. Any ideas?

**Comment by chrismlusk** [comment]
Re-reading the docs with fresh eyes, I see this note about the Success Behavior setting on the [Form Builder page](https://docs.solspace.com/craft/freeform/v3/overview/form-builder.html#settings-tabs):

> Success Behavior - set how you'd like the success return of this form to be handled. **May also be overrided at template-level.**

But I cannot find anything in the docs that describes how this can be overridden. What am I missing?

**Comment by kjmartens** [comment, marked as answer]
Hi @chrismlusk,

Please see my response to the other question here:
https://github.com/solspace/craft-freeform/discussions/465#discussioncomment-3766194

If you continue to need assistance with this issue you initially mentioned, please let me know. 🙂

**Reply by chrismlusk** [reply]
Hi, @kjmartens. Your response to #465 helps clarify that we are unable to override the actual `successBehaviour` property. Thank you for clearing that up.

As for this original issue, I'm curious if there are any potential updates to the plugin's AJAX submit method that could help. Maybe an option that I could set in a `freeform-ready` event handler? (Like the `skipHtmlReload` option, for example.)

**Reply by gustavs-gutmanis** [reply]
Hi @chrismlusk 

Could you please pick up the `dev-master` version and see if it works now?

**Reply by chrismlusk** [reply]
Hi @gustavs-gutmanis. Yes, adding the success events solves my issue. Thanks for the change!

**Reply by chrismlusk** [reply]
I'm sorry, I should have specified I'm still using v3. I saw the latest release for v4 included this change, so I just wanted to check if a release for v3 was also in the works. Thanks again.

**Reply by kjmartens** [reply]
@chrismlusk Yes, this will be in the next v3 release of Freeform as well 🙂

**Reply by chrismlusk** [reply]
@kjmartens Thank you. Any idea when the next v3 release might happen? This is a blocker for a feature update on our end.

**Reply by kjmartens** [reply]
We'll have it released later today. Thanks for your patience. 🙂
